### ISSUE TITLE
Linter provides no warnings for invalid C code

### ISSUE THREAD

**Issue by abbioro** [microsoft/vscode-cpptools#828]
See title, create a fresh `test.c` file with the contents:

```c
#include <stdio.h>

int main()
{
    asdfd;
    return 0;
}
```

(If) Linting is working properly, `asdfd` will be underlined with squigglies as you would expect. What you would not expect is that turning it into a function call makes the squigglies go away:

```c
#include <stdio.h>

int main()
{
    asdfd();
    return 0;
}
```

Furthermore, hovering over it produces the popup `int asdfd()`, as if it was defined to return int somewhere. This is just silly.

Related to #746

**Comment by bobbrow** [comment]
Indeed.  That is silly.  We'll investigate.

**Comment by sean-mcmanus** [comment]
This only repros in C files, not C++. It looks like the IntelliSense compiler is using the old default int function return value C89 behavior, so asdfd() is being interpreted as a declaration for "int asdfd()", even though we're passing in the C11 flag.

**Comment by abbioro** [comment]
Ok so it's valid C89, but surely this should be an error even in C89?

```c
#include <stdio.h>
#include <stdlib.h>

int main()
{
    char *line = maloc(100);
    return 0;
}
```

The misspelled malloc line shows squigglies only when placed *outside* the `main` function, but no squigglies inside. Still shows the int hint.

**Comment by sean-mcmanus** [comment]
Visual Studio 2017 also shows maloc (or any other undefined function call identifier) as being a function returning an int. I'll ask the VS-side people if they think our shared compiler has a bug.

**Comment by sean-mcmanus** [comment]
While it's not valid in the C99 standard, it looks like all the C compilers we support don't support this rule, I assume because it would be a breaking change with C89. So this is currently "by design", but if you can locate us a compiler (or linter) that enforces the non-implicit int behavior, we theoretically could add a new flag to trigger the enforcement behavior.

**Comment by abbioro** [comment]
Sorry I'm not sure I understand, the last example I posted doesn't compile with either GCC or Clang even with `-std=c89`, yet VS Code doesn't show any error squigglies for it. Is it not an error? Something is definitely wrong here. There should at least be warnings.

```
test.c:8:11: warning: incompatible integer to pointer conversion initializing 'char *' with an expression of type 'int' [-Wint-conversion]
    char *line = maloc(100);
          ^      ~~~~~~~~~~
1 warning generated.
/tmp/test-3f53c3.o: In function `main':
test.c:(.text+0x17): undefined reference to `maloc'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by sean-mcmanus** [comment]
Thanks for the info. The compiler is not giving you an error -- it successfully compiles and fails during linking. IntelliSense currently only gives compiler errors. This sounds like a feature request for warnings and/or linker errors.

**Comment by abbioro** [comment]
Warnings + linker errors sounds good to me.

**Comment by sean-mcmanus** [comment]
FYI, you can enable the warnings "experimental/hidden" feature via modifiying the msvc.json file next to the binary by removing the --no_warnings flag. Not sure when we'll get around to adding a setting for that. VS 2017 doesn't have these warnings enabled so it hasn't been tested as much. A previous issue reported was https://github.com/Microsoft/vscode-cpptools/issues/2091 .

![image](https://user-images.githubusercontent.com/19859882/43560325-c78fb548-95c6-11e8-9fe9-cc815fdd8aa8.png)

**Comment by alexanderkutin** [comment]
> FYI, you can enable the warnings "experimental/hidden" feature via modifiying the msvc.json file next to the binary by removing the --no_warnings flag. Not sure when we'll get around to adding a setting for that. VS 2017 doesn't have these warnings enabled so it hasn't been tested as much. A previous issue reported was #2091 .
> 
> ![image](https://user-images.githubusercontent.com/19859882/43560325-c78fb548-95c6-11e8-9fe9-cc815fdd8aa8.png)

Is it possible to enable "experimental/hidden" feature in Linux?

**Comment by sean-mcmanus** [comment]
@alexanderkutin Yes, the previous comment I made is outdated -- the file to modify is now common.json (not msvc.json).

**Comment by esotericist** [comment]
it seems that compilers generally support `-Werror-implicit-function-declaration` to produce errors instead of warnings for implicit functions. is there a way to get this parameter passed to the compiler for the intellisense behavior?

**Comment by sean-mcmanus** [comment]
@esotericist The warnings haven't been added to IntelliSense, but they'll be available via clang-tidy, but we're still working on the feature, so we don't know yet when it'll be available (see https://github.com/microsoft/vscode-cpptools/issues/2908):
![image](https://user-images.githubusercontent.com/19859882/135699969-47df2137-7b4f-4a1a-a4c2-1f857ea55e8d.png)

**Comment by esotericist** [comment]
i think my question was misunderstood. i'm not asking about warning support in vs code.

it is possible to tell the compiler to treat them as errors, not warnings, without treating all warnings as errors.

i can do this with my build setup, but i was wanting to know if we could pass that argument through vs code so they're treated as errors (not warnings) in the UI.

**Comment by sean-mcmanus** [comment]
@esotericist The `-Werror-implicit-function-declaration` is a clang or gcc flag, which our IntelliSense compiler doesn't process. You'd need to use the flag with your compiler or clang-tidy, and then it'll switch it from a warning to an error:
![image](https://user-images.githubusercontent.com/19859882/136089397-ddb8872d-0036-4c79-8330-21025c86a349.png)

**Comment by Heath123** [comment]
> FYI, you can enable the warnings "experimental/hidden" feature via modifiying the msvc.json file next to the binary by removing the --no_warnings flag. Not sure when we'll get around to adding a setting for that. VS 2017 doesn't have these warnings enabled so it hasn't been tested as much. A previous issue reported was #2091 .
> 
> ![image](https://user-images.githubusercontent.com/19859882/43560325-c78fb548-95c6-11e8-9fe9-cc815fdd8aa8.png)

What about if I'm using GCC on Linux?

**Comment by sean-mcmanus** [comment]
Linter errors for invalid C code can be shown now via clang-tidy (Right-click "Run Code Analysis on Active File", C_Cpp.codeAnalysis.clangTidy.enabled, etc.):

![image](https://user-images.githubusercontent.com/19859882/177613671-7ab0636f-38ef-4bbc-b6a8-62404f2310f0.png)

**Comment by cD1rtX3** [comment]
> While it's not valid in the C99 standard, it looks like all the C compilers we support don't support this rule, I assume because it would be a breaking change with C89.

In GCC, implicit function declaration is an error even without any additional warning flags. This is with `--std=c23`; I'm not sure about C89. You have to explicitly turn this off with `-Wno-implicit-function-declaration`.

**Comment by sean-mcmanus** [comment]
@cD1rtX3 Thanks for letting us know. Our EDG parser doesn't implement that yet (the MSVC compiler also doesn't), but we can file an issue to track that.
